docs/tutorials/javascript/amm/trade-with-auction-slot.md

```diff
@@ -0,0 +1,649 @@
+        // Tokens always have 15 significant digits;
+        // XRP is precise to integer drops, which can be as high as 10^17
+
+        let asset_out_bn = null
+        let pool_in_bn = null
+        let pool_out_bn = null
+        let isAmmAsset1Xrp = false
+        let isAmmAsset2Xrp = false
+
+        if ( takerPays.currency == 'XRP' ) {
+            asset_out_bn = BigNumber(xrpl.xrpToDrops(takerPays.amount)).precision(17)
+        } else {
+            asset_out_bn = BigNumber(takerPays.amount).precision(15)
+        }
+
+        if ( takerGets.currency == 'XRP' && asset1_currency == 'XRP' ) {
+            pool_in_bn = BigNumber(pool_asset1).precision(17)
+            isAmmAsset1Xrp = true
+        } else if ( takerGets.currency == 'XRP' && asset2_currency == 'XRP' ) {
+            pool_in_bn = BigNumber(pool_asset2).precision(17)
+            isAmmAsset2Xrp = true
+        } else if ( takerGets.currency == asset1_currency ) {
+            pool_in_bn = BigNumber(pool_asset1.value).precision(15)
+        } else {
+            pool_in_bn = BigNumber(pool_asset2.value).precision(15)
+        }
+
+        if (takerPays.currency == 'XRP' && asset1_currency == 'XRP' ) {
+            pool_out_bn = BigNumber(pool_asset1).precision(17)
+        } else if ( takerPays.currency == 'XRP' && asset2_currency == 'XRP' ) {
+            pool_out_bn = BigNumber(pool_asset2).precision(17)
+        } else if ( takerPays.currency == asset1_currency ) {
+            pool_out_bn = BigNumber(pool_asset1.value).precision(15)
+        } else {
+            pool_out_bn = BigNumber(pool_asset2.value).precision(15)
+        }
+
+        if ( takerPays.currency == 'XRP' && parseFloat(takerPays.amount) > parseFloat(xrpl.dropsToXrp(pool_out_bn)) ) {
+            results += `\n\nRequested ${takerPays.amount} ${takerPays.currency}, but AMM only holds ${xrpl.dropsToXrp(pool_out_bn)}. Quitting.`
+            standbyResultField.value = results
+            client.disconnect()
+            return
+        } else if ( parseFloat(takerPays.amount) > parseFloat(pool_out_bn) ) {
+            results += `\n\nRequested ${takerPays.amount} ${takerPays.currency}, but AMM only holds ${pool_out_bn}. Quitting.`
+            standbyResultField.value = results
+            client.disconnect()
+            return
+        }
+```
+
+Use [AMM helper functions](https://github.com/XRPLF/rippled/blob/2d1854f354ff8bb2b5671fd51252c5acd837c433/src/ripple/app/misc/impl/AMMHelpers.cpp) to estimate values for:
+
+- Cost to swap token without an auction slot.
+- Cost to swap token with an auction slot.
+- LP tokens to win an auction slot. This value factors the increase in the minimum bid of having new LP tokens issued to you from your deposit.
+- The amount of tokens for single-asset deposits to get the required LP tokens to win the auction slot.
+
+```javascript
+        // Use AMM's SwapOut formula to figure out how much of the takerGets asset
```

Comment:
As in the earlier tutorials, I'd prefer to see lengthy inline comments turned into descriptive text during the code walkthrough, so that the description is clearly separated from the code itself. It helps to reduce cognitive load.